cmd/kube-scheduler/app/options/options.go

```diff
@@ -186,6 +188,11 @@ func (o *Options) ApplyTo(c *schedulerappconfig.Config) error {
 			return err
 		}
 
+		// flag feature gates take precedence over feature gates in config file
+		for name, value := range o.ComponentConfig.FeatureGates {
+			cfg.FeatureGates[name] = value
+		}
+
 		// use the loaded config file only, with the exception of --address and --port. This means that
```

Comment:
with this change, this comment is no longer accurate. what is the use case for letting CLI feature flags override the config file? that makes a config file pretty difficult to reason about.

Reply:
I don't expect providers to set feature gates both in the config file AND the flag.

But I would expect providers to use a component config file, but setting feature gates only through the flag. This overriding should simplify adoption of component config and it's consistent with what other binaries are doing (such as kube).

Also see https://github.com/kubernetes/kubernetes/issues/84922#issuecomment-584255604

Reply:
This is an agreed status quo at the moment: https://github.com/kubernetes/kubernetes/issues/84922#issuecomment-577269381 

Other components already do it. It would be confusing to cluster administrators if scheduler behaved differently.

Reply:
This has been contentious every time it has come up, and the current state makes for hard-to-understand configurations. We cannot validate a standalone configuration file without knowing the full scheduler invocation. A configuration file author cannot understand the resulting behavior without also having the full invocation.

Reply:
The "config" flag help needs to be changed... it currently says "The path to the configuration file. Flags override values in this file" but CLI flags do not actually override most of the file values

Reply:
One way or other it will always be the case as long as the same options can be set in two different places.

What is a viable alternative here?

Reply:
> We cannot validate a standalone configuration file without knowing the full scheduler invocation

for example, should the flags be merged and set before calling ValidateKubeSchedulerConfiguration?

Reply:
> One way or other it will always be the case as long as the same options can be set in two different places.

If I'm reading the current code correctly, almost all the flags bound to fields that can be specified via config file are ignored when a config file is provided (we load the config file, then override `c.ComponentConfig = *cfg` with the loaded config, erasing all the values were bound to CLI flags

Reply:
It should say something along the lines: "flags not marked as deprecated override values in this file"

Reply:
Regarding the validation, one on line 187 checks if the `KubeSchedulerConfiguration` from the loaded config file is correct.

The validation of the `KubeSchedulerConfiguration` constructed using the flags is done before in `server.go`.

Finally, feature gates are validated when setting in `utilfeature.DefaultMutableFeatureGate`. It can and should be done earlier.  We're planning to have a separate PR to add feature flags validation in component-base which would allow to validate the feature gates before setting them and use in `KubeSchedulerConfiguration`.

Reply:
Yeah, it doesn't look like we make use of feature gates in validation currently, which is good (we should avoid introducing that in the future if possible).

We should update the flag help to reflect the actual behavior.

I still think the CLI flag overlay is more confusing than helpful (I haven't seen compelling use cases in which that was better than just setting the desired configuration in the config file), but as long as we clearly document and test the behavior, I won't block it. Should the config file field documentation call out that it can be overridden via a CLI flag?

Reply:
What @liggitt is saying is that we can NOT have such validation as long as the piecewise override from CLI flag happens. Such validation would be inconsistent with the validation done in `SetFromMap`